byteps/common/cpu_reducer.cc

```diff
@@ -56,171 +56,285 @@ bool CpuReducer::isRoot() {
+  for (size_t i = 0; i < (size_t)len; ++i) {
+    float in1_float;
+    float in2_float;
+    float out_float;
+    HalfBits2Float(in1 + i, &in1_float);
+    HalfBits2Float(in2 + i, &in2_float);
+    out_float = in1_float + in2_float * alpha;
     Float2HalfBits(&out_float, out + i);
   }
+#endif
+  return 0;
+}
+
+int CpuReducer::copy(void* dst, const void* src, size_t len) {
+//   auto in = reinterpret_cast<const float*>(src);
+//   auto out = reinterpret_cast<float*>(dst);
+// #pragma omp parallel for simd num_threads(_num_threads)
+//   for (size_t i = 0; i < len / 4; ++i) {
+//     out[i] = in[i];
+//   }
+//   if (len % 4) {
+//     std::memcpy(out + len / 4, in + len / 4, len % 4);
+//   }
+//   return 0;
+  std::memcpy(dst, src, len);
+}
+
+int CpuReducer::sign(void* dst, const void* src, size_t len, DataType dtype) {
+  switch (dtype) {
+    case BYTEPS_FLOAT32:
+      return _sign(reinterpret_cast<int32_t*>(dst),
+                   reinterpret_cast<const float*>(src), len);
+    case BYTEPS_FLOAT64:
+      return _sign(reinterpret_cast<int64_t*>(dst),
+                   reinterpret_cast<const double*>(src), len);
+    case BYTEPS_FLOAT16:
+      return _sign(reinterpret_cast<int16_t*>(dst),
+                   reinterpret_cast<const unsigned short*>(src), len);
+    default:
+      BPS_CHECK(0) << "Unsupported data type: " << dtype;
+  }
   return 0;
 }
 
-int CpuReducer::copy(void* dst, void* src, size_t len) {
-  auto in = (float*)src;
-  auto out = (float*)dst;
+template <typename T1, typename T2>
+size_t CpuReducer::_sign(T1* dst, const T2* src, size_t len) {
+  static_assert(sizeof(T1) == sizeof(T2), "T1 should be the same size as T2");
+
 #pragma omp parallel for simd num_threads(_num_threads)
-  for (size_t i = 0; i < len / 4; ++i) {
-    out[i] = in[i];
+  for (size_t i = 0; i < len / sizeof(T1); ++i) {
+    dst[i] = src[i] < 0;
   }
-  if (len % 4) {
-    std::memcpy(out + len / 4, in + len / 4, len % 4);
+
+  return len / sizeof(T1);
+}
+
+float CpuReducer::norm1(const void* src, size_t len, DataType dtype) {
+  switch (dtype) {
+    case BYTEPS_FLOAT32:
+      return _norm1(reinterpret_cast<const float*>(src), len);
+    case BYTEPS_FLOAT64:
+      return _norm1(reinterpret_cast<const double*>(src), len);
+    case BYTEPS_FLOAT16:
+      return _norm1_float16(src, len);
+    default:
+      BPS_CHECK(0) << "Unsupported data type: " << dtype;
   }
   return 0;
 }
 
+template <typename T>
+float CpuReducer::_norm1(const T* src, size_t len) {
+  float ret = 0;
+  // maybe use std::reduce in the future
+
+#pragma omp parallel for simd num_threads(_num_threads) reduction(+ : ret)
+  for (size_t i = 0; i < len / sizeof(T); ++i) {
+    ret += std::fabs(src[i]);
+  }
+
+  return ret;
+}
+
+float CpuReducer::_norm1_float16(const void* src, size_t len) {
+  // cast src and dst to your float16 type
+  auto in = reinterpret_cast<const unsigned short*>(src);
+  len = len / (size_t)2;
+
+  // #if __AVX512F__ && __F16C__
+  //    // TODO
```

Comment:
?

Reply:
i found some helpful instructions but they are in avx512. i don't remember the instructions now...   But actually this piece of code is unused. i will delete it.